eng/restore-toolset.sh

```diff
@@ -68,4 +68,52 @@ export NUGET_PACKAGES=$NUGET_PACKAGES
   echo "$scriptContents" > ${scriptPath}
 }
 
+# ReadVersionFromJson [json key]
+function ReadGlobalVersion {
+  local key=$1
+
+  if command -v jq &> /dev/null; then
+    _ReadGlobalVersion="$(jq -r ".[] | select(has(\"$key\")) | .\"$key\"" "$global_json_file")"
+  elif [[ "$(cat "$global_json_file")" =~ \"$key\"[[:space:]\:]*\"([^\"]+) ]]; then
+    _ReadGlobalVersion=${BASH_REMATCH[1]}
+  fi
+
+  if [[ -z "$_ReadGlobalVersion" ]]; then
+    Write-PipelineTelemetryError -category 'Build' "Error: Cannot find \"$key\" in $global_json_file"
+    ExitWithExitCode 1
+  fi
+}
+
+function CleanOutStage0ToolsetsAndRuntimes {
+  ReadGlobalVersion "dotnet"
+  local dotnetSdkVersion=$_ReadGlobalVersion
+  local dotnetRoot=$DOTNET_INSTALL_DIR
+  local versionPath="$dotnetRoot/.version"
+  local majorVersion="${dotnetSdkVersion:0:1}"
+  local aspnetRuntimePath="$dotnetRoot/shared/Microsoft.AspNetCore.App/$majorVersion.*"
+  local coreRuntimePath="$dotnetRoot/shared/Microsoft.NETCore.App/$majorVersion.*"
+  local wdRuntimePath="$dotnetRoot/shared/Microsoft.WindowsDesktop.App/$majorVersion.*"
+  local sdkPath="$dotnetRoot/sdk/$majorVersion.*"
+
+  if [ -f "$versionPath" ]; then
+    local lastInstalledSDK=$(cat $versionPath)
+    if [[ "$lastInstalledSDK" != "$dotnetSdkVersion" ]]; then
+      echo $dotnetSdkVersion > $versionPath
+      rm -rf $aspnetRuntimePath
+      rm -rf $coreRuntimePath
+      rm -rf $wdRuntimePath
+      rm -rf $sdkPath
+      rm -rf "$dotnetRoot/packs"
+      rm -rf "$dotnetRoot/sdk-manifests"
+      rm -rf "$dotnetRoot/templates"
+      Write-PipelineTelemetryError -category 'Build' "Found old version of SDK, cleaning out folder. Please run build.sh again"
+      ExitWithExitCode 1
+    fi
+  else
+    echo $dotnetSdkVersion > $versionPath
+  fi
+}
+
 InitializeCustomSDKToolset
+
+CleanOutStage0ToolsetsAndRuntimes
```

Comment:
General Q: I guess the design of Arcade requires duplicating scripts between 2 different scripting languages? I find that odd since modern PowerShell works cross-plat. I guess there must be some reason we couldn't just require PowerShell on Linux?

Reply:
That's a dnceng question. This is how it works across all repos. It might be a limitation of powershell on some linux distros.